### PR TITLE
[SPARK-37826][SQL] Use zstd codec name in ORC file names for hive orc impl

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Use the extensions map for ORC compression codecs in OrcUtils

### Why are the changes needed?

After SPARK-34954, we add zstd codec name in ORC file names for native orc impl w/o hive impl support.

This PR makes them consistent.

### Does this PR introduce _any_ user-facing change?

yes, with `spark.sql.orc.impl=hive`, you got the same file suffix as `=orc` for zstd compression

### How was this patch tested?

promote the related test to superclass for HiveOrcSourceSuite to trigger

### CODE REVIEW

**Review by LuciferYang — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM. Thank you, @yaooqinn and @LuciferYang .
Merged to master for Apache Spark 3.3.

**Comment by dongjoon-hyun** [comment]
cc @williamhyun , too

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Ur, I'm taking back my review. Did you test this manually, @yaooqinn ? It looks like causing a regression with `BasicWriteTaskStatsTracker`. Could you take a look at it?

```
scala> spark.range(3).write.option("compression", "zstd").orc("/tmp/after")
22/01/06 10:57:38 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
```

**Comment by yaooqinn** [comment]
> spark.range(3).write.option("compression", "zstd").orc("/tmp/after")

Yes, I have tested this PR with the spark-sql shell while sending this PR, and the result is correct.

I modified and verified the case you provided above w/ and w/o this PR, the warning always shows and not impact on the correctness. So I guess it's not related to this PR.

Currently, my guess is SingleDirectoryDataWriter call `newOutputWriter()` during `init` and `write` twice, and the later seems to try to visit a non-existent file in the former. I will take a look later.

**Comment by dongjoon-hyun** [comment]
So, do you mean it's a regression at Apache Spark 3.3 (on master) branch because Apache Spark 3.2.0 doesn't have this issue.
```
$ bin/spark-shell spark.sql.orc.impl=hive
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://localhost:4040
Spark context available as 'sc' (master = local[*], app id = local-1641533783429).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.0
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_312)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(3).write.option("compression", "zstd").orc("/tmp/after2")

scala> spark.version
res1: String = 3.2.0
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@yaooqinn . This is the current master result. I don't see any warning like that. Could you elaborate that how you can see the warnings on `master` branch, please?
```
$ bin/spark-shell spark.sql.orc.impl=hive
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/01/06 21:42:18 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://localhost:4040
Spark context available as 'sc' (master = local[*], app id = local-1641534138764).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.0-SNAPSHOT
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_312)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(3).write.option("compression", "zstd").orc("/tmp/master")

scala> spark.version
res1: String = 3.3.0-SNAPSHOT
```

**Comment by yaooqinn** [comment]
@dongjoon-hyun the logger here in 3.2 is `INFO`

```scala
 bin/spark-shell -c spark.sql.orc.impl=hive

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://10.242.189.214:4040
Spark context available as 'sc' (master = local[*], app id = local-1641535524296).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.0
      /_/

Using Scala version 2.12.15 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_251)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(3).write.option("compression", "zstd").orc("./spark-warehouse/zstdtest0")

scala> spark.read.text("./sparksql.log").filter(r => r.getString(0).contains("This could be due to the output format not")).head
res1: org.apache.spark.sql.Row = [22/01/07 14:05:37.195 Executor task launch worker for task 0.0 in stage 0.0 (TID 0) INFO BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.]

scala>
```

**Comment by dongjoon-hyun** [comment]
@yaooqinn . What do you think about `3.3.0-SNAPSHOT` result, https://github.com/apache/spark/pull/35117#pullrequestreview-846296996 ?

**Comment by dongjoon-hyun** [comment]
To be clear, you told me you saw the warning in the master branch like the following. However, I cannot see it.
> I modified and verified the case you provided above w/ and w/o this PR, the warning always shows and does not impact the correctness.

**Comment by yaooqinn** [comment]
> e you saw the warning in the maste

```
bin/spark-shell -c spark.sql.orc.impl=hive
22/01/07 14:22:27 WARN Utils: Your hostname, hulk.local resolves to a loopback address: 127.0.0.1; using 10.242.189.214 instead (on interface en0)
22/01/07 14:22:27 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/01/07 14:22:31 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://10.242.189.214:4040
Spark context available as 'sc' (master = local[*], app id = local-1641536551893).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.0-SNAPSHOT
      /_/

Using Scala version 2.12.15 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_251)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_1")
22/01/07 14:22:48 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.

scala> spark.version
res1: String = 3.3.0-SNAPSHOT

scala> spark.sql("select version()").show
+--------------------+
|           version()|
+--------------------+
|3.3.0 5880df41f50...|
+--------------------+


scala>
```

I have a pre-built version against master on Dec. 2021.

the case I use is `spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_1")`

you can help double-check on your side

**Comment by yaooqinn** [comment]
```
scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_2")
22/01/07 14:26:36 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.

scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_3")
22/01/07 14:26:39 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.

scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_4")
22/01/07 14:26:41 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.

scala> spark.range(0, 3, 1, 4).write.option("compression", "zstd").orc("./spark-warehouse/zstd_5")
22/01/07 14:26:44 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
```

w/ more rounds

**Comment by yaooqinn** [comment]
> bin/spark-shell spark.sql.orc.impl=hive

you missed `-c` before `spark.sql.orc.impl=hive`

**Comment by dongjoon-hyun** [comment]
Oh, got it. Let me try again.

**Comment by dongjoon-hyun** [comment]
Now, I verified the same situation in the master. Thanks.
```
$ bin/spark-shell -c spark.sql.orc.impl=hive
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/01/06 22:35:39 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://localhost:4040
Spark context available as 'sc' (master = local[*], app id = local-1641537340893).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.0-SNAPSHOT
      /_/

Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 1.8.0_312)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.range(3).write.option("compression", "zstd").orc("/tmp/zstd2")
22/01/06 22:36:01 WARN BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
```

Okay, we can leave it to another JIRA.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for your patience, @yaooqinn . And, thanks, @LuciferYang .
Merged to master for Apache Spark 3.3.

**Comment by yaooqinn** [comment]
thank you @dongjoon-hyun
